### PR TITLE
 Add discr_index to multi-variant layouts

### PR DESCRIPTION
**Pull request by tmandry**
We remove the assumption that the discriminant is always field 0, in
preparations for layouts like generators where this is not (always) going to be
the case.

Specifically, upvars are going to go before the discriminant. In theory, it's possible to remove _that_ assumption instead and keep the discriminant at field index 0, but one assumption or the other had to go :)

There is one place I know of in the debuginfo code where we'll still need to remove assumptions that the discriminant is the _only_ field. I was planning on doing this along with the upcoming generator change, which will also include tests that exercise the code changing in this PR.

r? @eddyb 
cc @oli-obk
cc @cramertj

### CODE REVIEW

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 1]
```diff
@@ -1504,7 +1507,8 @@ impl VariantMemberDescriptionFactory<'ll, 'tcx> {
 
 #[derive(Copy, Clone)]
 enum EnumDiscriminantInfo<'ll> {
-    RegularDiscriminant(&'ll DIType),
+    /// First field is the field index of the discriminant.
```
Comment (oli-obk) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: Please make this a struct variant with the documentation on the fields and not the variant

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 2]
```diff
@@ -1504,7 +1507,8 @@ impl VariantMemberDescriptionFactory<'ll, 'tcx> {
...
+    RegularDiscriminant(usize, &'ll DIType),
```
Comment (oli-obk) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: Use https://github.com/rust-lang/rust/blob/1d4f0e71c6cd55d43f1d7b9b58cab8ef7c6f4117/src/librustc/mir/mod.rs#L2004 as the type instead of `usize` and adjust all use sites to match if possible (instead of using `usize` or `u32`). The latter can also be a separate PR, since it's a preexisting issue. You can use the `as_u32` and `as_usize` methods on `Field` for now.

File: src/librustc_target/abi/mod.rs [hunk 3]
```diff
@@ -828,12 +828,13 @@ pub enum Variants {
...
-    /// Enums with more than one inhabited variant: for each case there is
-    /// a struct, and they all have space reserved for the discriminant,
-    /// which is the sole field of the enum layout.
+    /// Enum-likes with more than one inhabited variant: for each case there is
+    /// a struct, and they all have space reserved for the discriminant.
+    /// For enums this is the sole field of the layout.
     Multiple {
         discr: Scalar,
         discr_kind: DiscriminantKind,
+        discr_index: usize,
```
Comment (oli-obk) [inline src/librustc_target/abi/mod.rs]: also use https://github.com/rust-lang/rust/blob/1d4f0e71c6cd55d43f1d7b9b58cab8ef7c6f4117/src/librustc/mir/mod.rs#L2004 here

**Review by oli-obk — COMMENTED** [review]
What's the motivation for preferring this PR over making the generator layouts have field `0` always be the discriminant? Our fields do not have to be ordered by memory order, so the generator layouting code could always reorder the fields so that the discriminant comes first.

**Comment by tmandry** [comment]
> What's the motivation for preferring this PR over making the generator layouts have field 0 always be the discriminant? Our fields do not have to be ordered by memory order, so the generator layouting code could always reorder the fields so that the discriminant comes first.

That's another possibility, but would have required another refactoring in other parts of the code that assume the "discriminant" is a field other than 0. I wasn't sure which approach would be better, maybe @eddyb knows?

**Comment by oli-obk** [comment]
> would have required another refactoring in other parts of the code that assume the "discriminant" is a field other than 0

I'm confused. Isn't this what your PR is doing and the alternative is that no code outside the generator layout computation code would have to change?

**Comment by tmandry** [comment]
Sorry for not being more clear. A lot of the code that handles generators and/or _closures of any kind_ assumes that the upvars start at field index 0. So AIUI, I would have to adjust all the lowering code to MIR, as well as how they are handled in debuginfo.

It doesn't seem obvious to me that that would be an easier or less invasive change, but then I really don't know.

File: src/librustc_target/abi/mod.rs (hunk 3 above)
Comment (eddyb) [reply src/librustc_target/abi/mod.rs]: We're in `rustc_target`. Also, this whole module uses `usize` (if not `u32`) for field indices.

**Comment by eddyb** [comment]
@oli-obk So what we want to get is basically a builtin version of this:
```rust
enum Foo {
    a: A,
    b: B,
    Bar(X),
    Baz(Y),
}
```

Those shared fields, `a` and `b`, *already* have "semantic" indices `0` and `1`, and we wouldn't want to have to do ±1 everywhere to handle them - we used to have this with enum variants (because each variant layout had a copy of the discriminant at the start) and it was really annoying to work with (since it's "an off by one error by default").

In the case of generators, the upvars already have indices, so we have the same problem, with fields we don't want to complicate ourselves with renumbering.

<hr/>

An alternative to including the discriminant field index would be to say "the last field is the discriminant" instead of "field 0 is the discriminant".
Do we think that is an improvement over this PR?

File: src/librustc_target/abi/mod.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc_target/abi/mod.rs]: yea, this doesn't have to happen in this PR, and we can move the `Field` definition up into rustc_target

**Comment by oli-obk** [comment]
Thanks, that motivation makes total sense.

I think encoding the last field would be very confusing. Having an explicit discriminant field index is the best solution indeed, I was just missing the context.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs (hunk 2 above)
Comment (tmandry) [reply src/librustc_codegen_llvm/debuginfo/metadata.rs]: Done.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs (hunk 1 above)
Comment (tmandry) [reply src/librustc_codegen_llvm/debuginfo/metadata.rs]: Done

**Review by oli-obk — APPROVED** [review]

**Comment by cramertj** [comment]
@eddyb Sorry if i'm missing something, but this example here:
```
enum Foo {
    a: A,
    b: B,
    Bar(X),
    Baz(Y),
}
```
wouldn't be forwards-compatible with having some fields which were valid for multiple variants but not for *all* variants, which we want to support at some point, right? Is this just an artifact of how you wrote the syntax, or is this actually a limitation present in the design at this point?

**Comment by tmandry** [comment]
@cramertj We haven't discussed how to do this yet. I'm thinking we can either shoehorn it into this scheme (multiple vairants that have the same field at the same offset), or introduce an entirely new concept in the layout.

Either way, this still seems like a good first step to me.

**Comment by tmandry** [comment]
So @cramertj and I were talking about this some more. One approach that might be simpler in the long run is, rather than putting always-valid fields on the outer layout, we put them in the same place on every variant. Something like this:

```rust
enum Foo {
  Unresumed(A, B),
  YieldedOnce(A, B, X),
  YieldedTwice(A, B, Y),
}
```

Then we can implicitly leave those fields in place as we cycle through variants. It seems like this would be helpful later, when/if we want to optimize fields that are there across more than one variant.

What do people think about this? If it makes sense and it doesn't add considerable complexity to do it now, maybe we should go that route instead.

**Comment by eddyb** [comment]
@cramertj @tmandry In that example, the common fields are upvars *not* always-valid locals.
What @tmandry just said is what I expected we were talking about all along, wrt locals across yields.

**Comment by tmandry** [comment]
Okay, I was conflating those two things for some reason.

Would we ever want to do liveness analysis of non-Drop upvars and recover space from those as well?

In any case I think this PR is well-motivated and ready to merge.

**Review by eddyb — APPROVED** [review]

**Comment by tmandry** [comment]
Rebased to make a dependent PR easier.

@bors r=eddyb
